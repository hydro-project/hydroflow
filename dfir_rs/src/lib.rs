#![warn(missing_docs)]

//! Hydroflow is a low-level dataflow-based runtime system for the [Hydro Project](https://hydro.run/).
//!
//! The primary item in this crate is the [`Hydroflow`](crate::scheduled::graph::Dfir) struct,
//! representing a Hydroflow dataflow graph. Although this graph can be manually constructed, the
//! easiest way to instantiate a `Hydroflow` instance is with the [`dfir_syntax!`] macro using
//! Hydroflow's custom "surface syntax."
//!
//! ```rust
//! let mut hf = dfir_rs::dfir_syntax! {
//!     source_iter(["hello", "world"]) -> for_each(|s| println!("{}", s));
//! };
//! hf.run_available();
//! ```
//!
//! For more examples, check out the [`examples` folder on Github](https://github.com/hydro-project/hydroflow/tree/main/hydroflow/examples).

pub mod compiled;
pub mod scheduled;
pub mod util;

#[cfg(feature = "meta")]
pub use dfir_lang as lang;
#[cfg(feature = "python")]
pub use pyo3;
pub use variadics::{self, var_args, var_expr, var_type};
pub use {
    bincode, bytes, futures, itertools, lattices, pusherator, rustc_hash, serde, serde_json, tokio,
    tokio_stream, tokio_util, tracing, web_time,
};

/// `#[macro_use]` automagically brings the declarative macro export to the crate-level.
mod declarative_macro;
#[cfg(feature = "dfir_datalog")]
pub use dfir_datalog::*;
#[cfg(feature = "dfir_macro")]
pub use dfir_macro::{
    dfir_main as main, dfir_parser, dfir_syntax, dfir_syntax_noemit, dfir_test as test,
    monotonic_fn, morphism, DemuxEnum,
};

// TODO(mingwei): Use the [nightly "never" type `!`](https://doc.rust-lang.org/std/primitive.never.html)
/// Stand-in for the [nightly "never" type `!`](https://doc.rust-lang.org/std/primitive.never.html)
pub type Never = std::convert::Infallible;

#[cfg(doctest)]
mod booktest {
    mod surface_ops {
        dfir_macro::surface_booktest_operators!();
    }
}
