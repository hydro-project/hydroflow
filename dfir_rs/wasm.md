# WASM

## Tests

### Setup

Install `wasm-bindgen-cli` globally: `cargo install wasm-bindgen-cli`.
`wasm-bindgen-cli` provides a test runner harness.

### Running Tests

Run (in the root hydroflow directory):

```
CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER=wasm-bindgen-test-runner cargo test --target wasm32-unknown-unknown -p hydroflow --tests
```

Flag explanation:

- `CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER=wasm-bindgen-test-runner`: Tells
  Cargo to use the test harness provided by `wasm-bindgen-cli`.
- `-p hydroflow`: Only run tests in the hydroflow directory.
- `--tests`: Only run tests; do not build examples. Many (possibly all) of the
  examples in hydroflow/examples currently do not compile to WASM because they
  use networking.


### Adding a new WASM test

Instead of the `#[test]` attribute, mark the test with the
`#[multiplaform_test]` attribute.

```rust
use multiplatform_test::multiplatform_test;

#[multiplatform_test]
fn my_test() {
  // ...
}
```
