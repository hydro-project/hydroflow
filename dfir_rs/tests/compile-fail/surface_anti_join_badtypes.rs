use dfir_rs::dfir_syntax;

fn main() {
    let mut df = dfir_syntax! {
        j = anti_join() -> for_each(std::mem::drop);
        source_iter(0..10) -> map(|x| (x, x)) -> [pos]j;
        source_iter(0..10) -> map(|_| "string") -> [neg]j;
    };
    df.run_available();
}
