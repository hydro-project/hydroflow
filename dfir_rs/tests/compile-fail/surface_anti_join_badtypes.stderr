error[E0271]: type mismatch resolving `<ForEach<{closure@surface_anti_join_badtypes.rs:4:18}, {integer}> as Pusherator>::Item == &str`
 --> tests/compile-fail/surface_anti_join_badtypes.rs:5:13
  |
5 |           j = anti_join() -> for_each(std::mem::drop);
  |               -^^^^^^^^^^
  |               |
  |  _____________expected `&str`, found integer
  | |
6 | |         source_iter(0..10) -> map(|x| (x, x)) -> [pos]j;
7 | |         source_iter(0..10) -> map(|_| "string") -> [neg]j;
  | |_______________________________________________- required by a bound introduced by this call
  |
note: required by a bound in `pivot_run_sg_2v1`
 --> tests/compile-fail/surface_anti_join_badtypes.rs:5:13
  |
5 |           j = anti_join() -> for_each(std::mem::drop);
  |  _____________^
6 | |         source_iter(0..10) -> map(|x| (x, x)) -> [pos]j;
7 | |         source_iter(0..10) -> map(|_| "string") -> [neg]j;
  | |_______________________________________________^ required by this bound in `pivot_run_sg_2v1`
